internal/validator/validator.go

```diff
@@ -0,0 +1,202 @@
+
+func (v Validator) String() string {
+	return "validator"
+}
+
+func (v Validator) Start(ctx context.Context, ready chan<- struct{}) error {
+	// create and attempt to insert the first epoch
+	epoch := Epoch{
+		StartBlock: v.inputBoxDeploymentBlock,
+		EndBlock:   v.inputBoxDeploymentBlock + v.epochDuration,
+	}
+	if err := v.repo.InsertFirstEpochTransaction(ctx, epoch); err != nil {
+		return err
+	}
+	ready <- struct{}{}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			// add a backoff strategy for the actual implementation
+
+			// the current epoch might be the one we just created or another
+			// if this isn't the first run
+			currentEpoch, err := v.repo.GetCurrentEpoch(ctx)
+			if err != nil {
+				return err
+			}
+			latestBlock, err := v.repo.GetMostRecentBlock(ctx)
+			if err != nil {
+				return err
+			}
+			// test if should finish epoch
+			if latestBlock >= currentEpoch.EndBlock {
+				outputs, err := v.repo.GetAllOutputsFromProcessedInputs(
+					ctx,
+					currentEpoch.StartBlock,
+					currentEpoch.EndBlock,
+					nil,
+				)
+				if err != nil {
+					return err
+				}
+				newEpoch := Epoch{
+					StartBlock: currentEpoch.EndBlock + 1,
+					EndBlock:   currentEpoch.EndBlock + v.epochDuration,
+				}
+				if len(outputs) == 0 {
+					err := v.repo.FinishEmptyEpochTransaction(ctx, newEpoch)
+					if err != nil {
+						return err
+					}
+				} else {
+					lastIndex := len(outputs) - 1
+					inputRange := InputRange{
+						First: outputs[0].InputIndex,
+						Last:  outputs[lastIndex].InputIndex,
+					}
+					machineStateHash, err := v.repo.GetMachineStateHash(ctx, inputRange.Last)
+					if err != nil {
+						return err
+					}
+					proofs, outputsEpochRootHash, err := v.generateProofs(
+						ctx,
+						inputRange,
+						machineStateHash,
+						outputs,
+					)
+					if err != nil {
+						return err
+					}
+
+					epochHash := crypto.Keccak256(outputsEpochRootHash, machineStateHash)
+					if err = v.repo.FinishEpochTransaction(
+						ctx,
+						newEpoch,
+						&Claim{
+							InputRange: inputRange,
+							EpochHash:  epochHash,
+						},
+						proofs,
+					); err != nil {
+						return err
+					}
+				}
+			}
+		}
+	}
+}
+
+// ------------------------------------------------------------------------------------------------
+// Validator functions
+// ------------------------------------------------------------------------------------------------
+
+func NewValidator(
+	repo ValidatorRepository,
+	epochDuration uint64,
+	inputBoxDeploymentBlock uint64,
+) Validator {
+	return Validator{repo, epochDuration, inputBoxDeploymentBlock}
+}
+
+// generateProofs will create the proofs for all Outputs within an Epoch.
+// It returns the proofs and the root hash of the Merkle tree used to generate them
+func (v Validator) generateProofs(
+	ctx context.Context,
+	inputRange InputRange,
+	machineStateHash hexutil.Bytes,
+	outputs []Output,
+) ([]Proof, hexutil.Bytes, error) {
```

Comment:
Do we need to return the OutputsEpochRootHash as a different value? Can't we get it from any of the proofs?

Reply:
We can. I'll remove it from the returned values.